kernel/qGraph.ml

```diff
@@ -0,0 +1,330 @@
+(************************************************************************)
+(*         *      The Rocq Prover / The Rocq Development Team           *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+open Sorts
+
+module ElimTable = struct
+  open Quality
+
+  let const_eliminates_to q q' =
+    match q, q' with
+    | QType, _ -> true
+    | QProp, (QProp | QSProp) -> true
+    | QSProp, QSProp -> true
+    | (QProp | QSProp), _ -> false
+
+  let eliminates_to q q' =
+    match q, q' with
+    | QConstant QType, _ -> true
+    | QConstant q, QConstant q' -> const_eliminates_to q q'
+    | QVar q, QVar q' -> QVar.equal q q'
+    | (QConstant _ | QVar _), _ -> false
+end
+
+module G = AcyclicGraph.Make(struct
+    type t = Quality.t
+    module Set = Quality.Set
+    module Map = Quality.Map
+
+    let equal = Quality.equal
+    let compare = Quality.compare
+
+    let raw_pr = Quality.raw_pr
+
+    let anomaly_err q = Pp.(str "Quality " ++ Quality.raw_pr q ++ str " undefined.")
+  end)
+
+module RigidPath = struct
+  type t = Quality.t * Quality.t
+
+  let compare (q1,q2) (q1',q2') =
+    let i = Quality.compare q1 q1' in
+    if i = 0 then Quality.compare q2 q2'
+    else i
+end
+
+module RigidPaths = Set.Make(RigidPath)
+
+module QMap = QVar.Map
+module QSet = QVar.Set
+
+type t =
+  { graph: G.t;
+    rigid_paths: RigidPaths.t;
+    ground_and_global_sorts: Quality.t list;
+    dominant: Quality.t QMap.t;
+    delayed_check: QSet.t QMap.t;
+  }
+
+type path_explanation = G.explanation Lazy.t
+
+type explanation =
+  | Path of path_explanation
+  | Other of Pp.t
+
+type quality_inconsistency =
+  ((QVar.t -> Pp.t) option) * (ElimConstraint.kind * Quality.t * Quality.t * explanation option)
+
+let cached_graph = ref None
+
+(* If s can eliminate to s', we want an edge between s and s'.
+   In the acyclic graph, it means setting s to be lower or equal than s'.
+   This function ensures a uniform behaviour between [check] and [enforce]. *)
+let to_graph_cstr k =
+  let open ElimConstraint in
+  match k with
+    | ElimTo -> AcyclicGraph.Le
+    | Equal -> AcyclicGraph.Eq
+
+let check_func k =
+  let open AcyclicGraph in
+  match to_graph_cstr k with
+  | Le -> G.check_leq
+  | Eq -> G.check_eq
+  | Lt -> assert false
```

Comment:
This assertion is useless, you could match directly on `k` instead of adding dead code.